SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -488,29 +489,95 @@ private int scheduleTasks(SingularityRequest request, RequestState state,
       }
 
       if (request.isRackSensitive() && configuration.isRebalanceRacksOnScaleDown()) {
-        List<SingularityTaskId> extraCleanedTasks = new ArrayList<>();
-        int numActiveRacks = rackManager.getNumActive();
-        double perRack = request.getInstancesSafe() / (double) numActiveRacks;
-
-        Multiset<String> countPerRack = HashMultiset.create();
-        for (SingularityTaskId taskId : remainingActiveTasks) {
-          countPerRack.add(taskId.getRackId());
-          LOG.info("{} - {} - {} - {}", countPerRack, perRack, extraCleanedTasks.size(), taskId);
-          if (countPerRack.count(taskId.getRackId()) > perRack && extraCleanedTasks.size() < numActiveRacks / 2) {
-            extraCleanedTasks.add(taskId);
-            LOG.info("Cleaning up task {} to evenly distribute tasks among racks", taskId);
-            taskManager.createTaskCleanup(new SingularityTaskCleanup(pendingRequest.getUser(), TaskCleanupType.REBALANCE_RACKS, now, taskId, Optional.<String>absent(), Optional.<String>absent(), Optional.<SingularityTaskShellCommandRequestId>absent()));
+        rebalanceRacks(request, state, deployStatistics, pendingRequest, maybePendingDeploy, now, remainingActiveTasks);
+      }
+      if (request.getSlaveAttributeMinimums().isPresent()) {
+        rebalanceAttributeDistribution(request, state, deployStatistics, pendingRequest, maybePendingDeploy, now, remainingActiveTasks);
+      }
+    }
+    return numMissingInstances;
+  }
+
+  private void rebalanceAttributeDistribution(
+      SingularityRequest request,
+      RequestState state,
+      SingularityDeployStatistics deployStatistics,
+      SingularityPendingRequest pendingRequest,
+      Optional<SingularityPendingDeploy> maybePendingDeploy,
+      long now,
+      List<SingularityTaskId> remainingActiveTasks) {
+    Map<String, Map<String, Integer>> instanceAttrCounts = new HashMap<>();
+
+    for (SingularityTaskId taskId: remainingActiveTasks) {
+      SingularitySlave slave = leaderCache.getSlave(taskManager.getTask(taskId).get().getMesosTask().getSlaveId().getValue()).get();
```

Comment:
Totally missed that slave manager already has a `if (leaderCache.active()) {return leaderCache.getSlave(slaveId);}`, my bad on that one. Guess we can use slaveManager after all in these spots